### ISSUE TITLE
bash script causes NVDA to freeze

### ISSUE THREAD

**Issue by mehgcap** [nvaccess/nvda#15787]
### Steps to reproduce:

First, let me say that this is an edge case and a low priority. It's also something that NVDA's maintainers are likely already aware of, if by other causes.

In short, a script I wrote manages to freeze NVDA, stopping all speech and braille output until I restart NVDA with the hotkey. I can't provide the script here, as it is meant to do work on a database I can't share. It also caused the problem *before* I fixed some bugs in the script, so the current version doesn't have the same problem. The NVDA freeze only showed up when things failed, and now that the script works, the problem can't be reproduced. I think the key, though, is that the script was outputting a lot of text very quickly. I plan to make a script that dumps a lot of text and does nothing else. If I can reproduce the problem that way, I'll add the script to this issue.

The script runs on Linux and is executed over SSH, which I access through cmd.exe on Windows. I haven't tried this with Microsoft's Terminal app yet. Once executed, the script examines a few dozen text files, analyzes each, and uses the data to generate a couple million SQL queries. It then executes these in sequence, pausing every few hundred to avoid overwhelming the server on which the queries execute.

For output, the script writes a line of text each time it moves to a new file. If all goes well, a single line below the file name is printed. This line uses print("string", "\r") (Python) to update itself. It gives a count of how many queries have run, how many there are total, and how many have failed. Using this print() command, the single line updates in real time.

I believe the problem comes when the script runs into an error. When that happens, it prints the exception message and the first thousand characters of the query to the console window. Until I fixed the problems and adjusted the logic, many queries would fail. Since the exceptions were caught and handled, the loop continued. That is, a query would fail, some 1300 characters would print, then the next query would fail, another 1300 or so characters would print, and so on. This would repeat multiple times per second.

### Actual behavior:

NVDA would simply freeze. Both speech and braille would stop working entirely. If I hit ctrl-win-n, NVDA would play the startup sound, but not the shutdown sound, indicating I was starting it and not restarting it. If I moved focus away from the cmd window, everything was fine. If I focused on the window and there were a lot of errors, NVDA would fail almost immediately.

### Expected behavior:

NVDA shouldn't lock up like this. I'm probably overwhelming something, but NVDA should handle the problem more gracefully, if possible. At worst, when this happens, perhaps NVDA could alert the user that they should close or switch away from the current application.

### NVDA logs, crash dumps and other attachments:
```
NVDA initialized
INFO - watchdog.waitForFreezeRecovery (15:18:50.254) - watchdog (17308):
Starting freeze recovery after 0.5003790000000006 seconds.
ERROR - watchdog.waitForFreezeRecovery (15:19:05.270) - watchdog (17308):
Core frozen in stack! (15.5157051 seconds)
INFO - watchdog.waitForFreezeRecovery (15:19:05.276) - watchdog (17308):
Listing stacks for Python threads:
Python stack for thread 19752 (LegacyWinConsole._monitorThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "NVDAObjects\behaviors.pyc", line 433, in _monitor

Python stack for thread 17308 (watchdog):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "watchdog.pyc", line 158, in _watcher
  File "watchdog.pyc", line 186, in waitForFreezeRecovery
  File "watchdog.pyc", line 65, in getFormattedStacksForAllThreads

Python stack for thread 10428 (Thread-1):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "C:\Users\mehgc\AppData\Roaming\nvda\addons\enhancedTones\globalPlugins\enhancedTones\_tones.py", line 227, in run
    self.waitBeep.wait()
  File "threading.pyc", line 552, in wait
  File "threading.pyc", line 296, in wait

Python stack for thread 9552 (piper4nvda_asyncio):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "C:\Users\mehgc\AppData\Roaming\nvda\addons\piper_neural_voices\synthDrivers\piper_neural_voices\aio.py", line 37, in _thread_target
    ASYNCIO_EVENT_LOOP.run_forever()
  File "C:\Users\mehgc\AppData\Roaming\nvda\addons\piper_neural_voices\synthDrivers\piper_neural_voices\lib\asyncio\base_events.py", line 585, in run_forever
    self._run_once()
  File "C:\Users\mehgc\AppData\Roaming\nvda\addons\piper_neural_voices\synthDrivers\piper_neural_voices\lib\asyncio\base_events.py", line 1921, in _run_once
    event_list = self._selector.select(timeout)
  File "selectors.pyc", line 323, in select
  File "selectors.pyc", line 314, in _select

Python stack for thread 2908 (globalPlugins.rdAccess.directoryChanges.DirectoryWatcher):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "hwIo\ioThread.pyc", line 283, in run

Python stack for thread 15608 (addons.rdAccess.lib.ioThreadEx.IoThreadEx):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "hwIo\ioThread.pyc", line 283, in run

Python stack for thread 21196 (winInputHook):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "winInputHook.pyc", line 81, in hookThreadFunc

Python stack for thread 16592 (UIAHandler.UIAHandler.MTAThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "UIAHandler\__init__.pyc", line 522, in MTAThreadFunc
  File "queue.pyc", line 170, in get
  File "threading.pyc", line 296, in wait

Python stack for thread 2884 (synthDrivers._espeak.BgThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "synthDrivers\_espeak.pyc", line 201, in run
  File "queue.pyc", line 170, in get
  File "threading.pyc", line 296, in wait

Python stack for thread 14328 (hwIo.ioThread.IoThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "hwIo\ioThread.pyc", line 283, in run

Python stack for thread 14864 (MainThread):
  File "nvda.pyw", line 399, in <module>
  File "core.pyc", line 814, in main
  File "wx\core.pyc", line 2237, in MainLoop
  File "wx\core.pyc", line 3407, in <lambda>
  File "core.pyc", line 747, in processRequest
  File "core.pyc", line 763, in Notify
  File "queueHandler.pyc", line 97, in pumpAll
  File "monkeyPatches\comtypesMonkeyPatches.pyc", line 130, in newCpbDel
  File "comtypes\__init__.pyc", line 957, in __del__
  File "comtypes\__init__.pyc", line 1212, in Release
  File "monkeyPatches\comtypesMonkeyPatches.pyc", line 32, in __call__
```

#### NVDA installed/portable/running from source:

installed

#### NVDA version:

2023.3

#### Windows version:

Version 22H2 (OS Build 19045.3693)

#### Name and version of other software in use when reproducing the issue:

cmd.exe

#### Does the issue still occur after restarting your computer?

yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

no

#### If NVDA add-ons are disabled, is your problem still occurring?

yes

**Comment by Adriani90** [comment]
Is this also reproducible with the last alpha version of nvda?

**Comment by XLTechie** [comment]
Is this a duplicate of #15786?

**Comment by XLTechie** [comment]
@mehgcap, while you say you had add-ons disabled, your log fragment tells a 
different story. It shows at least two of them active when the freeze happened. 
One is piper_neural_voices, which as far as I know is still in beta, but either 
way could introduce unpredictable factors.

Perhaps you should reconstruct the original circumstances, and capture it 
happening with a debug log with add-ons disabled?

As an aside, when a script can generate that much output, wouldn't it be better 
to capture it to a file instead, then review the file with less after the 
operation?

E.g. script 2>&1 > outfile

Or put the output-generating portion of the script in a brace construct, and 
have the brace construct's output captured in a file, which the script could 
then page to you at the end?

There are dozens of other ways to do this, in whole or in part, without spraying 
huge quantities of output to the terminal.

**Comment by mehgcap** [comment]
@XLTechie Yes, there are better ways. The script wasn't meant to work in the way that caused the NVDA freeze. Normally, the single, updating line of output for each file being processed worked fine and caused no problems. It was only when thousands of queries started failing in a row that I ran into the issue. As I said, I've since fixed the script, which is part of why I didn't put it here. Additionally, that there are better ways to output text in my specific case isn't the point. The freeze happened, seemingly because a lot of text was dumped to the window very quickly. That could happen for multiple reasons.

As to add-ons, I have Piper installed, but I'm not using it. My synthesizer is ESpeak-NG, the one included with NVDA. If I can reproduce the freeze with a different script, I'll grab a log with add-ons completely disabled. You're right that I should have put "not known" for that section.

It looks similar to #17586, but I haven't had any problem with my cmd window closing completely like that ticket mentions. It's only NVDA that's affected.

@Adriani90 Once I create the test script, I'll try the alpha and see if it's fixed there.

**Comment by seanbudd** [comment]
This may be fixed by https://github.com/nvaccess/nvda/pull/14888

**Comment by mehgcap** [comment]
@seanbudd I don't believe so. Still on the same version of NVDA, I ran the command in the notes (`git -P log -1000`). NVDA did not freeze or do anything unexpected. I tried with 100000 instead of 1000, and still had no problems. This was on a local repo of NVDA in case that matters.

**Comment by seanbudd** [comment]
@mehgcap - you reproduced the issue using a build from the draft code in #14888? Note this PR needs updating and is in a draft state still, but the intention of the PR is to fix issues like this.

**Comment by mehgcap** [comment]
No. I tried to reproduce the issue using the public release of NVDA by using the command, and was not able to do so. Thus, the change the PR solves doesn't seem to be quite the same.

**Comment by seanbudd** [comment]
Are you suggesting that this is fixed is now fixed 2023.3 or alpha?

**Comment by seanbudd** [comment]
are you testing now with piper voices / add-ons add-ons disabled? Note, even if eSpeak is selected, the add-on is still loaded. You can disable the add-on in the add-on store.

**Comment by Adriani90** [comment]
Given #14888 is merged into alpha and is meant to fix such issues, and also given the fact that the initial author cannot reproduce this issue anymore, I am closing as works for me. If you are still having this issue please comment and we can reopen.
